src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -113,40 +113,25 @@ class ModelQueue {
   ~ModelQueue() = default;
 
   void add_task(PredictTask task) {
-    std::unique_lock<std::mutex> l(queue_mutex_);
+    std::lock_guard<std::mutex> lock(queue_mutex_);
     Deadline deadline = std::chrono::system_clock::now() +
                         std::chrono::microseconds(task.latency_slo_micros_);
     queue_.emplace(deadline, std::move(task));
+    condition_.notify_one();
   }
 
   int get_size() {
     std::unique_lock<std::mutex> l(queue_mutex_);
     return queue_.size();
   }
 
-  // TODO: If we implement the scheduler so that it calls
-  // get_earliest_deadline(), then uses that to compute the batch size,
-  // then calls dequeue with that batch size, it's possible for a new
-  // task with an earlier deadline to be submitted. I'm not sure what the
-  // correct behavior here should be.
-  boost::optional<Deadline> get_earliest_deadline() {
-    std::unique_lock<std::mutex> l(queue_mutex_);
-    remove_tasks_with_elapsed_deadlines();
-    if (!queue_.empty()) {
-      return boost::optional<Deadline>(queue_.top().first);
-    } else {
-      return boost::optional<Deadline>{};
-    }
-  }
-
-  // Dequeues up to max_batch_size tasks from the queue without blocking
-  std::vector<PredictTask> get_batch(int max_batch_size) {
-    // NOTE: Because the queue lock is released and reacquired between a
-    // call to get_earliest_deadline() (which blocks until the queue is
-    // not empty) and the call to this method, it's possible for the queue
-    // to be empty, in which case the returned vector will have size 0.
-    std::unique_lock<std::mutex> l(queue_mutex_);
+  std::vector<PredictTask> get_batch(
+      std::function<int(Deadline)> &&get_batch_size) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
     remove_tasks_with_elapsed_deadlines();
+    condition_.wait(lock, [this]() { return !queue_.empty(); });
```

Comment:
We should call `remove_tasks_with_elapsed_deadlines` again after the condition is met and the lock is obtained because it's possible that expired tasks are in the queue at this stage.